<launch>
    <arg name="veh" doc="Name of vehicle. ex: megaman"/>
    <arg name="local" default="false" doc="true for running on laptop. false for running on vehicle."/>
    <arg name="config" default="baseline" doc="Specify a config."/>
    <arg name="param_file_name" default="default" doc="Specify a param file. ex:megaman." />
    <arg name="camera" default="true" doc="camera set to false means only apriltags no camera nodes are launched"/>
    <arg name="live" default="true" doc="live set to false means we don't run the actual camera (i.e. running from a log file" />

    <include file="$(find duckietown)/machines"/>

    <!-- position_controller -->
    <remap from="position_controller_node/lane_pose" to="lane_filter_node/lane_pose"/>
    <remap from="position_controller_node/apriltags_in" to="tag_detections"/>
    <node if="$(arg local)" ns="$(arg veh)" name="position_controller_node" pkg="platooning" type="position_controller.py" output="screen" />
    <node unless="$(arg local)" machine="$(arg veh)" ns="$(arg veh)" name="position_controller_node" pkg="platooning" type="position_controller.py" output="screen" />

       <!-- lane filter -->
    <include file="$(find duckietown)/launch/lane_filter.launch">
         <arg name="veh" value="$(arg veh)"/>
         <arg name="local" value="$(arg local)"/>
         <arg name="config" value="$(arg config)"/>
         <arg name="param_file_name" value="$(arg param_file_name)"/>
         <arg name="live" value="$(arg live)"/>
    </include>



    <!-- AprilTags Detections -->
    <remap from="image_rect" to="camera_node/image/compressed" />
    <remap from="camera_node/image/camera_info" to="camera_node/camera_info" />
    <include file="$(find apriltags_ros)/launch/apriltag_detector_node.launch">
        <arg name="veh" value="$(arg veh)"/>
        <arg name="local" value="$(arg local)"/>
        <arg name="config" value="$(arg config)"/>
        <arg name="param_file_name" value="$(arg param_file_name)"/>
    </include>


    <!-- run inverse_kinematics_node -->
    <remap from="inverse_kinematics_node/car_cmd" to="position_controller_node/car_cmd"/>
    <remap from="inverse_kinematics_node/wheels_cmd" to="wheels_driver_node/wheels_cmd" />
    <include file="$(find dagu_car)/launch/inverse_kinematics_node.launch">
        <arg name="veh" value="$(arg veh)"/>
        <arg name="local" value="$(arg local)"/>
        <arg name="config" value="$(arg config)"/>
    </include>

    <!-- run forward_kinematics_node -->
    <remap from="forward_kinematics_node/wheels_cmd" to="wheels_driver_node/wheels_cmd_executed" />
    <include file="$(find dagu_car)/launch/forward_kinematics_node.launch">
        <arg name="veh" value="$(arg veh)"/>
        <arg name="local" value="$(arg local)"/>
        <arg name="config" value="$(arg config)"/>
    </include>

    <!-- run the velocity_to_pose_node -->
    <remap from="velocity_to_pose_node/velocity" to="forward_kinematics_node/velocity" />
    <include file="$(find dagu_car)/launch/velocity_to_pose_node.launch">
        <arg name="veh" value="$(arg veh)"/>
        <arg name="local" value="$(arg local)"/>
        <arg name="config" value="$(arg config)"/>
    </include>

    <!-- Wheels Driver -->
    <include file="$(find dagu_car)/launch/wheels_driver_node.launch">
        <arg name="veh" value="$(arg veh)"/>
    </include>

    
</launch>

